example/Disparity.hs

```diff
@@ -0,0 +1,93 @@
+{-# LANGUAGE AllowAmbiguousTypes, CPP, DataKinds, FlexibleInstances, GADTs,
+             QuantifiedConstraints, RankNTypes, TypeFamilies #-}
+{-# OPTIONS_GHC -Wno-incomplete-patterns  #-}
+{-# OPTIONS_GHC -Wno-missing-export-lists #-}
+{-# OPTIONS_GHC -Wno-compat-unqualified-imports #-}
+{-# OPTIONS_GHC -Wno-missing-signatures #-}
+{-# OPTIONS_GHC -Wno-incomplete-uni-patterns #-}
+
+module Disparity where
+import Prelude
+import GHC.TypeLits
+import HordeAd.Core.DualNumber
+import HordeAd.Core.DualClass           as DC
+import HordeAd
+import qualified System.Random          as R
+import qualified Data.Array.ShapedS     as OS
+import qualified Foreign.Storable       as F
+import Data.List
+
+
+-- | Disparity cost volume.
+--
+--   Take two arrays of multi channel 2d images, where the first contains
+--   left views of the scene and the second contains right views.
+--
+--   For each pair of images, slice the right image over the left image,
+--   and for each offset produce the L1 distance indicating how well correponding
+--   multi-channel image elements in the right image match those in the left.
+--
+--   Described in:
+--    Anytime Stereo Image Depth Estimation on Mobile Devices
+--    Wang, Lai et al, ICRA 2019
+--    https://arxiv.org/abs/1810.11408
+--    Section III b).
+--
+costVolume
+ :: forall d r nImgs nChas nRows nCols nCount shA shO.
+    ( ADModeAndNum d r
+    , KnownNat nImgs, KnownNat nChas, KnownNat nRows, KnownNat nCols, KnownNat nCount
+    , shA ~ '[nImgs, nChas,  nRows, nCols]
+    , shO ~ '[nImgs, nCount, nRows, nCols])
+ => Int
+ -> StaticNat nCount
+ -> ADVal d (OS.Array shA r)
+ -> ADVal d (OS.Array shA r)
+ -> ADVal d (OS.Array shO r)
+
+costVolume iStart _sCount arrL arrR =
+  buildS @shO $ \[iImg, iDisp, iRow, iCol] ->
+    let arrVecL = buildS @'[nChas] $ \[iCha] ->
+                    indexzS0 arrL [iImg, iCha, iRow, iCol]
+        iSrc    = iCol - iStart - iDisp
+        arrVecR = buildS @'[nChas] $ \[iCha] ->
+                    indexzS0 arrR [iImg, iCha, iRow, iSrc]
+    in  sumElements0
+         $ flattenS1
+         $ zipWithOuterS (\xL xR -> abs (xL - xR)) arrVecL arrVecR
+
+
+
+-- Write out an example test vector.
+testCostVolume
+ = let  arrL    = random @'[1, 2, 4, 6] @Double 1
+        arrR    = random @'[1, 2, 4, 6] @Double 2
+        arrS    = random @'[1, 4, 4, 6] @Double 3
+        arrO    = primal $ costVolume 0 (MkSN :: StaticNat 4) (constant arrL) (constant arrR)
+        arrDL   = revDt (\aL -> costVolume 0 MkSN aL (constant arrR)) arrL arrO
+        arrDR   = revDt (\aR -> costVolume 0 MkSN (constant arrL) aR) arrR arrO
+   in   putStrLn $ unlines
+         [ "arrL  = " ++ show arrL
+         , "arrR  = " ++ show arrR
+         , "arrS  = " ++ show arrS
+         , "arrO  = " ++ show arrO
+         , "arrDR = " ++ show arrDR
+         , "arrDL = " ++ show arrDL ]
+
+
+-- | Generate an array of random values for testing.
+random
```

Comment:
This is a different implementation of

https://github.com/Mikolaj/horde-ad/blob/master/src/HordeAd/External/Adaptor.hs#L278

though yours may be more convenient by permitting type applications.

Reply:
I've just pushed a tiny change that probably makes `randomVals @'[1, 2, 4, 6] @Double` possible, just as in your reimplementation.